sale_order_type/models/account_move.py

```diff
@@ -32,10 +40,35 @@ def _compute_sale_type_id(self):
                 if sale_type:
                     record.sale_type_id = sale_type
 
-    @api.onchange("sale_type_id")
-    def onchange_sale_type_id(self):
-        # TODO: To be changed to computed stored readonly=False if possible in v14?
-        if self.sale_type_id.payment_term_id:
-            self.invoice_payment_term_id = self.sale_type_id.payment_term_id.id
-        if self.sale_type_id.journal_id:
-            self.journal_id = self.sale_type_id.journal_id.id
+    @api.depends("sale_type_id")
+    def _compute_invoice_payment_term_id(self):
+        if hasattr(super(), "_compute_invoice_payment_term_id"):
+            super()._compute_invoice_payment_term_id()
+        for rec in self:
+            if (
+                rec.sale_type_id
+                and rec.invoice_payment_term_id != rec.sale_type_id.payment_term_id
+            ):
+                rec.invoice_payment_term_id = rec.sale_type_id.payment_term_id
+
+    @api.depends("sale_type_id")
+    def _compute_journal_id(self):
+        if hasattr(super(), "_compute_journal_id"):
+            super()._compute_journal_id()
+        for rec in self:
+            if (
+                rec.sale_type_id.journal_id
+                and rec.journal_id != rec.sale_type_id.journal_id
+            ):
+                rec.journal_id = rec.sale_type_id.journal_id
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        res = super().create(vals_list)
+        # This will trigger recomputations and overload default values
+        # for the new computed fields, if set on the sale.order.type
+        res.modified(["sale_type_id"])
+        # For some reason journal_id is not following this rule
+        # and needs to be explicitly recomputed
+        res._compute_journal_id()
+        return res
```

Comment:
You can't launch `_compute_journal_id` on create. This way, you are overriding any specific `journal_id` passed in the vals, which is not correct. Having also `res.modifed...` seems not correct. Why do you need that? Please restore the previous situation and let me see the error in CI.

Reply:
Yeah, you're right.

This is what I've tried to solve:

> Apparently computed writeable stored fields are not recomputed on create if the field is explicitly set on vals, or if there's a default value.
 Tests were failing because of that, mainly:
warehouse_id is set explicitly here: https://github.com/odoo/odoo/blob/13.0/addons/sale_stock/models/sale_order.py#L66
picking_policy also has a default value:
https://github.com/odoo/odoo/blob/13.0/addons/sale_stock/models/sale_order.py#L27
journal_id has a default value:
https://github.com/odoo/odoo/blob/13.0/addons/account/models/account_move.py#L121
and others..
So they're not recomputed on record create.


I've just pushed a commit reverting this, so you can test.